File: 23. Ch-5-6.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.flag = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
        self.count = 0

    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        if not node.flag:
            node.flag = True
            self.count += 1

    def UniqueWords(self):
        return self.count

n = int(input())
trie = Trie()
for i in range(n):
    trie.insert(input())
print(trie.UniqueWords())

File: 23. Ch-5-6.java
import java.util.*;
class TrieNode 
{
    Map<Character, TrieNode> children;
    boolean flag;
    public TrieNode() 
    {
        children = new HashMap<>();
        flag = false;
    }
}

class Trie 
{
    TrieNode root= new TrieNode();
    int Count=0;

    void insert(String word) 
    {
        TrieNode current = root;
        for (int i = 0; i < word.length(); i++) 
        {
            char ch = word.charAt(i);
            TrieNode node = current.children.get(ch);
            if (node == null) 
            {
                node = new TrieNode();
                current.children.put(ch, node);
            }
            current = node;
        }

        if (!current.flag) 
        {
            current.flag = true;
            Count++;
        }
    }

    int UniqueWordCount() 
    {
        return Count;
    }
}

public class Main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        Trie trie = new Trie();
        int n = sc.nextInt();
        for (int i = 0; i < n; i++) 
        {
            String word = sc.next();
            trie.insert(word);
        }

        System.out.println(trie.UniqueWordCount());
    }
}
